File: src/main/kotlin/alg02deques/Deque.kt
package alg02deques

interface Deque<T> : Iterable<T> {
    val size: Int
    val isEmpty: Boolean
        get() = this.size == 0
    val indices: IntRange
        get() = IntRange(0, this.size - 1)

    operator fun get(index: Int): T
    operator fun set(index: Int, item: T)

    fun addFirst(item: T)
    fun addLast(item: T)
    fun addAt(index: Int, item: T)
    fun removeFirst(): T
    fun removeAt(index: Int): T
    fun removeLast(): T
}

File: alg02deques/deque.go
package alg02deques

type Deque[T any] interface {
	ToSlice() []T
	Size() int
	IsEmpty() bool
	Get(index int) T
	AddFirst(item T)
	AddLast(item T)
	AddAtIndex(index int, item T)
	RemoveFirst() T
	RemoveLast() T
	RemoveAtIndex(index int) T
}

func isEmpty[T any](deque Deque[T]) bool {
	return deque.Size() == 0
}

func toSlice[T any](deque Deque[T]) []T {
	slice := make([]T, deque.Size())
	for index := 0; index < deque.Size(); index++ {
		slice[index] = deque.Get(index)
	}
	return slice
}
